### PR TITLE
[SPARK-47833][SQL][CORE] Supply caller stackstrace for checkAndGlobPathIfNecessary AnalysisException

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

SPARK-29089 parallelized `checkAndGlobPathIfNecessary` by leveraging ForkJoinPool, it also introduced a side effect, if something goes wrong, the reported error message loses caller side stack trace. 

For example, I meet the following error in a Spark job, I have no idea what happened without the caller stack trace.

```
2024-04-12 14:31:21 CST ApplicationMaster INFO - Final app status: FAILED, exitCode: 15, (reason: User class threw exception: org.apache.spark.sql.AnalysisException: Path does not exist: hdfs://xyz-cluster/user/abc/hive_db/tmp.db/tmp_lskkh_1
	at org.apache.spark.sql.errors.QueryCompilationErrors$.dataPathNotExistError(QueryCompilationErrors.scala:1011)
	at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$4(DataSource.scala:785)
	at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$4$adapted(DataSource.scala:782)
	at org.apache.spark.util.ThreadUtils$.$anonfun$parmap$2(ThreadUtils.scala:372)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
)
```

### Why are the changes needed?

Improve error message.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT is added, and the exception stacktrace differences are

raw stacktrace
```
java.lang.RuntimeException: Error occurred on Thread-9
        at org.apache.spark.util.ThreadUtilsSuite$$anon$3.internalMethod(ThreadUtilsSuite.scala:141)
        at org.apache.spark.util.ThreadUtilsSuite$$anon$3.run(ThreadUtilsSuite.scala:138)
```

enhanced exception stacktrace
```
java.lang.RuntimeException: Error occurred on Thread-9
        at org.apache.spark.util.ThreadUtilsSuite$$anon$3.internalMethod(ThreadUtilsSuite.scala:141)
        at org.apache.spark.util.ThreadUtilsSuite$$anon$3.run(ThreadUtilsSuite.scala:138)
        at ... run in separate thread: Thread-9 ... ()
        at org.apache.spark.util.ThreadUtilsSuite.$anonfun$new$16(ThreadUtilsSuite.scala:151)
        at org.scalatest.enablers.Timed$$anon$1.timeoutAfter(Timed.scala:127)
        at org.scalatest.concurrent.TimeLimits$.failAfterImpl(TimeLimits.scala:282)
        at org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:231)
        at org.scalatest.concurrent.TimeLimits.failAfter$(TimeLimits.scala:230)
        at org.apache.spark.SparkFunSuite.failAfter(SparkFunSuite.scala:69)
        at org.apache.spark.SparkFunSuite.$anonfun$test$2(SparkFunSuite.scala:155)
        at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
        at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
        (... other scalatest callsites)
```

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by pan3793** [comment]
cc @srowen @viirya @dongjoon-hyun @yaooqinn

**Comment by pan3793** [comment]
cc @gengliangwang @LuciferYang @mridulm WDYT of this approach for stacktrace enhancement? Or do you have other suggestions?

**Review by mridulm — COMMENTED** [review]
Looks reasonable to me, but I will let @viirya and @dongjoon-hyun comment better.

Btw, are there other callsites where this might be helpful ?

**Comment by pan3793** [comment]
@mridulm In this case, `ThreadUtils.parmap` catches and wraps the internal `AnalysisException` with a `SparkException`, if we throw `SparkException` directly, the callsites are retained, but propagating `SparkException` instead of the real exception `AnalysisException` might not desirable behavior, I believe the original author also wants to keep `AnalysisException` so unwrap in-place. I checked other callers of `ThreadUtils.parmap`, they throw `SparkException` directly so no such issues.

File: core/src/main/scala/org/apache/spark/util/ThreadUtils.scala [hunk 1]
```diff
@@ -261,31 +261,50 @@ private[spark] object ThreadUtils {
...
+       realException: T,
+       combineMessage: String =
+         s"run in separate thread using ${ThreadUtils.getClass.getName.stripSuffix("$")}",
+       dropStacks: Int = 1): T = {
+    require(dropStacks >= 0, "dropStacks must be zero or positive")
+    // Remove the part of the stack that shows method calls into this helper method
+    // This means drop everything from the top until the stack element
+    // ThreadUtils.wrapCallerStack(), and then drop that as well (hence the `drop(1)`).
+    // Large dropStacks allows caller to drop more stacks.
+    val baseStackTrace = Thread.currentThread().getStackTrace
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/util/ThreadUtils.scala:290]:
```scala
    val simpleName = this.getClass.getSimpleName
    ...
    val baseStackTrace = Thread.currentThread().getStackTrace
      .dropWhile(!_.getClassName.contains(simpleName))
      .drop(dropStacks)

    // Remove the part of the new thread stack that shows methods call from this helper method
    val extraStackTrace = realException.getStackTrace
      .takeWhile(!_.getClassName.contains(simpleName))
```

Comment (pan3793) [reply core/src/main/scala/org/apache/spark/util/ThreadUtils.scala:291]: changed as suggested

File: core/src/main/scala/org/apache/spark/util/ThreadUtils.scala [hunk 2]
```diff
@@ -261,31 +261,50 @@ private[spark] object ThreadUtils {
...
-        throw realException
+        throw wrapCallerStacktrace(realException, dropStacks = 2)
       case None =>
         result
     }
   }
 
+  /**
+   * Adjust exception stack stace to wrap with caller side thread stack trace.
+   * The exception stack traces will be like the following
```
[the remark this answers is not in the document]
Comment (pan3793) [reply core/src/main/scala/org/apache/spark/util/ThreadUtils.scala]: addressed

File: core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala [hunk 3]
```diff
@@ -128,6 +128,42 @@ class ThreadUtilsSuite extends SparkFunSuite {
...
+      }
+    }
+    t.setDaemon(true)
+    t.setUncaughtExceptionHandler { case (_, e) => exception = e }
+    t.start()
+    t.join()
+
+    ThreadUtils.wrapCallerStacktrace(exception, s"run in separate thread: $runnerThreadName")
+
+    assert(exception.getStackTrace.mkString("\n").contains(
```
Comment (yaooqinn) [inline core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala:150]: nit: val stStr = exception.getStackTrace.mkString("\n")

File: core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala [hunk 4]
```diff
@@ -128,6 +128,42 @@ class ThreadUtilsSuite extends SparkFunSuite {
...
+    )
+    assert(exception.getStackTrace.mkString("\n").contains(
+      s"... run in separate thread: $runnerThreadName ..."),
+      "stack trace does not contain expected place holder"
+    )
+    assert(exception.getStackTrace.mkString("\n").contains(
+      "org.scalatest.Suite.run"),
+      "stack trace does not contain caller stack trace"
+    )
+    assert(exception.getStackTrace.mkString("\n").contains("ThreadUtils.scala") === false,
```
Comment (yaooqinn) [inline core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala:162]: nit: !exception.getStackTrace.mkString("\n").contains("ThreadUtils.scala")

File: core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala [hunk 5]
```diff
@@ -128,6 +128,42 @@ class ThreadUtilsSuite extends SparkFunSuite {
     )
   }
 
+  test("wrapCallerStacktrace") {
```
Comment (yaooqinn) [inline core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala:131]: attach jira id

File: core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala (hunk 3 above)
Comment (pan3793) [reply core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala]: changed as suggested

File: core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala (hunk 4 above)
Comment (pan3793) [reply core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala]: changed as suggested

File: core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala (hunk 5 above)
Comment (pan3793) [reply core/src/test/scala/org/apache/spark/util/ThreadUtilsSuite.scala]: added

**Comment by pan3793** [comment]
@yaooqinn thanks for reviewing, I addressed all comments

**Review by yaooqinn — APPROVED** [review]
LGTM, thank you for updating the desc

**Comment by yaooqinn** [comment]
Merged to master. Thank you @pan3793

**Pull request closed without merging**